src/login/logind-dbus.c

```diff
@@ -1916,6 +1916,8 @@ static int method_do_shutdown_or_sleep(
                 a = handle_action_lookup(HANDLE_KEXEC);
         else if ((flags & SD_LOGIND_SOFT_REBOOT))
                 a = handle_action_lookup(HANDLE_SOFT_REBOOT);
+        else if (kexec_loaded())
+                a = handle_action_lookup(HANDLE_KEXEC);
```

Comment:
the first branch does the same, this should be combined.

maybe add a switch SD_LOGIND_REBOOT_NOT_VIA_KEXEC to turn off this automatic "upgrade"?

Reply:
SD_LOGIND_REBOOT_VIA_KEXEC as no effect after this, afaics.

Reply:
so maybe do it like this: just set the existing flag `SD_LOGIND_REBOOT_VIA_KEXEC` in the client, and do not change the server?

Reply:
and then do the soft reboot similar. add a new flag that enables this new "upgrade" behaviour (maybe called: SD_LOGIND_SOFT_REBOOT_IF_SET_UP or so), and allow it to be set together via the existing SD_LOGIND_REBOOT_VIA_KEXEC mode. And in that case, do a soft reboot if /run/nextboot exists, and if not check for a kexec kernel and use it if there is, otherwise do regular reboot. Then change systemctl to set both flags by default.

hence there would be:

1. SD_LOGIND_SOFT_REBOOT → do a soft reboot no matter what
2. SD_LOGIND_SOFT_REBOOT_IF_SET_UP → do a soft reboot if /run/nextroot/ is set up
3. SD_LOGIND_REBOOT_VIA_KEXEC → do a kexec if kernel is loaded for kexec
4. SD_LOGIND_SOFT_REBOOT_IF_SET_UP|SD_LOGIND_REBOOT_VIA_KEXEC → if /run/nextroot/ exists do that, otherwise check kexec and do that, otherwise do regular reboot.

Then change systemctl to default to the latter flag combo.

Reply:
ok